README.md
# STE-C_55X_Autopilot
A simulation of the S-TEC Fifty Five X Autopilot as a plugin for X-Plane 11

Please read the manual for further information


STE-C 55X Autopilot/changelog.txt
#######################################################################################################################################################################################################
																				STE-C 55X Autopilot Plugin	-	Changelog
#######################################################################################################################################################################################################
v0.5.1-beta (current)
-----------------------------------------------------------
-	investigated 'turnoff on CDI change' using the GTN
	-> see manual 'Quicktipp/FAQ' section for help

-	added trim annunciation as described in POH
	-> DISCLAIMER: Yoke trimswitch being ignored due to XPlane limitation
	-> also see settings to for new option according to this

v0.5.0-beta
-----------------------------------------------------------
-	added settings tab
		-	you can now select whether you want to start with or without the AP showing up
		-	added legal advice
		
-	added license tab to see license in Sim
-	added changelog tab to see changelog


v0.4.0-beta
-----------------------------------------------------------
-	initial public beta release


v0.3.0-alpha
-----------------------------------------------------------
-	initial alpha release for private alpha test

STE-C 55X Autopilot/data/modules/Custom Module/settings.lua
--[[
STE-C 55X Autopilot Plugin, a plugin to bring the "S-TEC FIFTY FIVE X" into your XPlane Flight Sim.
Copyright (C) 2022 Luka Emanuel Nöckel

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

   	You should have received a copy of the GNU General Public License
   	along with this program.  If not, see <https://www.gnu.org/licenses/>.

You can contact me via E-MAIL: quackderbruchpilot05@gmail.com, via GitHub (QuackDerBruchpilot) or the X-Plane.org forum (@flightsimmer1710)

	ANY USE OF THIS CODE CONSTITUTES ACCEPTANCE OF THE TERMS OF THE COPYRIGHT NOTICE.
	The current software uses the SASL library. Redistribution in source and binary forms, with 
	or without modification, are not permitted.

	Copyright (c) 2020, 1-sim.aero
	All rights reserved.

	SASL copyright holders are not responsible in any event for malfunction, unavailability, 
	incompatibility or other problems that might occur while using the end product which uses 
	SASL in conjunction with any other software. SASL copyright holders are not liable for any 
	properties for any end product which uses SASL.
]]--

addSearchPath(moduleDirectory.."/Custom Module/")												-- loading settings
include("config")																				--

btn_save_color  = {0.153, 0.204, 0.255}										-- standard save button color
btn_licenses_color = {0.153, 0.204, 0.255}									-- standard licences button color
btn_changelog_color = {0.153, 0.204, 0.255}									-- standard licences button color

addSearchResourcesPath(moduleDirectory.."/images/")												-- add resources directory
gantari = sasl.gl.loadFont("Gantari-SemiBold.ttf")												-- define font





function onMouseDown(window, x, y, button, parentX, parentY)							-- mouse click event handler
	if button == MB_LEFT then
		if 3 <= x and x <= 23 and 143 <= y and y <= 163 then								-- checkbox 1
			AP_wnd_on_startup = not AP_wnd_on_startup											-- switch state for AP window on satrtup
		
			return 1
		end
		
		if 3 <= x and x <= 23 and 123 <= y and y <= 143 then								-- checkbox 2
			has_elecTrim = not has_elecTrim														-- switch state for has_elecTrim
			
			return 1
		end
		
		if 323 <= x and x <= 383 and 100 <= y and y <= 124 then								-- save button
		
			btn_save_color = {0.204, 0.229, 0.255}
			
			local settings = io.open(moduleDirectory.."/Custom Module/config.lua", "w+")		-- load config file
			io.output(settings)																	--
			io.write("first_launch = false\n")													-- save that we already saw the settings window once
			
			io.write("AP_wnd_on_startup = "..tostring(AP_wnd_on_startup).."\n")					-- save AP_wnd_on_startup
			
			io.write("has_elecTrim = ".. tostring(has_elecTrim).."\n")							-- save has_elecTrim setting
			io.close(settings)																	-- close file
			
			return 1
		end
		
		if 50 <= x and x <= 130 and 20 <= y and y <= 50 then								-- see licenses button
			
			btn_licenses_color = {0.204, 0.229, 0.255}
			show_license()
			return 1
		end
		
		if 270 <= x and x <= 350 and 20 <= y and y <= 50 then								-- see changelog button
			
			btn_changelog_color = {0.204, 0.229, 0.255}
			include("changelog")
			return 1
		end
	end
end





function wnd_changelog_close()
	sasl.logInfo("Changelog window closed")
end

function show_license()																	-- licenses text box
	sasl.messageWindow	(
		500, 500, 500, 500, "STE-C 55X Autopilot License",
		"STE-C 55X Autopilot Plugin, a plugin to bring the 'S-TEC FIFTY FIVE X' into your XPlane Flight Sim. \n \n Copyright (C) 2022 Luka Emanuel Nöckel \n \n This program is free software: You can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or	(at your option) any later version. This program is distributed in the hope that it will be useful, but without any warranty; without even the implied warranty of merchantability or fitness for a particular purpose. See the GNU General Public License for more details. \n You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>. \n \n You can contact me via E-MAIL: quackderbruchpilot05@gmail.com, via GitHub (QuackDerBruchpilot) or the X-Plane.org forum (@QuackDerBruchpilot05) \n \n \n \n ANY USE OF THIS CODE CONSTITUTES ACCEPTANCE OF THE TERMS OF THE COPYRIGHT NOTICE. \n The current software uses the SASL library. Redistribution in source and binary forms, with or without modification, are not permitted.	Copyright (c) 2020, 1-sim.aero	All rights reserved. SASL copyright holders are not responsible in any event for malfunction, unavailability, incompatibility or other problems that might occur while using the end product which uses	SASL in conjunction with any other software. SASL copyright holders are not liable for any properties for any end product which uses SASL.",
		1,"Close", wnd_license_close
	)
end

function wnd_license_close()
	sasl.logInfo("License Window closed")
end





function onMouseUp(window, x, y, button, parentX, parentY)								-- mouse release event handler
	if button == MB_LEFT then
		if 323 <= x and x <= 383 and 100 <= y and y <= 124 then							
			btn_save_color = {0.153, 0.204, 0.255}											-- change color of save button back to standard
			return 1
		end
		
		if 50 <= x and x <= 130 and 20 <= y and y <= 50 then								-- change color of license button back to standard
			btn_licenses_color = {0.153, 0.204, 0.255}
			return 1
		end
		
		if 270 <= x and x <= 350 and 20 <= y and y <= 50 then								-- change color of changelog button back to standard
			btn_changelog_color = {0.153, 0.204, 0.255}
			return 1
		end
	end
end





function update()																		-- things we have to do every cycle

	if AP_wnd_on_startup then 																-- Decide if Checkboxes should be checked or not
		chk_AP_wnd_on_startup = sasl.gl.loadImage("icons8-checked-checkbox-20.png")			--
	else																					--
		chk_AP_wnd_on_startup = sasl.gl.loadImage("icons8-unchecked-checkbox-20.png")		--
	end																						--
	
	if has_elecTrim then																	--
		chk_has_elecTrim = sasl.gl.loadImage("icons8-checked-checkbox-20.png")				--
	else																					--
		chk_has_elecTrim = sasl.gl.loadImage("icons8-unchecked-checkbox-20.png")			--
	end																						--
end





local function text(x, y, txt)
	sasl.gl.drawText(gantari, x, y, txt, 12, false, false, TEXT_ALIGN_LEFT, {1.0, 1.0, 1.0})
end

local function text_warning(x, y, txt)
	sasl.gl.drawText(gantari, x, y, txt, 12, true, false, TEXT_ALIGN_LEFT, {1.0, 0.0, 0.0})
end





function draw()																			-- draw things into the settings window

	text(4, 285, "STE-C 55X Autopilot Plugin - Copyright (C) 2022 Luka Emanuel Nöckel")			-- copyright owner
	sasl.gl.drawLine(0, 277, 400, 277, {210, 210, 210})											-- spacer
	
	text(3, 263, "This program comes with ABSOLUTELY NO WARRANTY. This is free")				--
	text(3, 250, "software, and you are welcome to redistribute it under certain")				-- copyright notice
	text(3, 237, "conditions. For details open the License file in the plugin's directory.")	--
	
	text_warning(45, 215, "NO GUARENTEED REAL BEHAVIOR! FOR FLIGHT")							-- legal advice
	text_warning(100, 200, "SIMULATION PURPOSE ONLY!")											-- please take seriously
	sasl.gl.drawLine(0, 185, 400, 185, {210, 210, 210})											-- spacer
	
	text(3, 170, "Settings:")														-- settings section
	
		sasl.gl.drawTexture(chk_AP_wnd_on_startup, 3, 143, 20, 20)								-- draw checkbox for AP_wnd_on_startup
		text(35, 150, "open autopilot window on startup")										-- text for checkbox
		
		sasl.gl.drawTexture(chk_has_elecTrim, 3, 123, 20, 20)									-- for has_elecTrim
		text(35, 130, "connect autopilot with electric trim")									--
		
	sasl.gl.drawRectangle(323, 100, 60, 20, btn_save_color)										-- Save button
	text(340, 106, "Save")																		--
	
	sasl.gl.drawRectangle(50, 20, 80, 30, btn_licenses_color)									-- licences button
	text(56, 32, "see licences")																--
	
	sasl.gl.drawRectangle(270, 20, 80, 30, btn_changelog_color)									-- changelog button
	text(280, 32, "changelog")																	--
end






sasl.logInfo("settings window loaded")










STE-C 55X Autopilot/data/modules/Custom Module/changelog.lua
--[[
STE-C 55X Autopilot Plugin, a plugin to bring the "S-TEC FIFTY FIVE X" into your XPlane Flight Sim.
Copyright (C) 2022 Luka Emanuel Nöckel

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

   	You should have received a copy of the GNU General Public License
   	along with this program.  If not, see <https://www.gnu.org/licenses/>.

You can contact me via E-MAIL: quackderbruchpilot05@gmail.com, via GitHub (QuackDerBruchpilot) or the X-Plane.org forum (@flightsimmer1710)

	ANY USE OF THIS CODE CONSTITUTES ACCEPTANCE OF THE TERMS OF THE COPYRIGHT NOTICE.
	The current software uses the SASL library. Redistribution in source and binary forms, with 
	or without modification, are not permitted.

	Copyright (c) 2020, 1-sim.aero
	All rights reserved.

	SASL copyright holders are not responsible in any event for malfunction, unavailability, 
	incompatibility or other problems that might occur while using the end product which uses 
	SASL in conjunction with any other software. SASL copyright holders are not liable for any 
	properties for any end product which uses SASL.
]]--

sasl.messageWindow	(
	500, 500, 500, 500, "STE-C 55X Autopilot Changelog",
		
	"v0.5.1-beta (current) \n ----------------------------------------------------------- \n -	investigated 'turnoff on CDI change' \n 		-	see manual 'quicktipp/FAQ' section \n \n -	implemented trim annunciation as described in POH \n 		-	DISCLAIMER: Yoke trimswitch is being ignored because of XPlane limitation \n 		-	see manual for new settings option \n \n \n v0.5.0-beta \n ----------------------------------------------------------- \n -	added settings tab \n 		-	you can now select whether you want to start with or without the AP           showing up \n \n		-	added legal advice \n \n -	added license tab to see license in Sim \n -	added changelog tab to see changelog \n \n \n v0.4.0-beta \n ----------------------------------------------------------- \n -	initial public beta release \n \n \n v0.3.0-alpha \n ----------------------------------------------------------- \n -	initial alpha release for private alpha test",
		
	1,"Close", wnd_changelog_close
)


-- -	implemented trim annunciation as described in POH \n		-	Disclaimer: Yoke Trimswitch is being ignored because of XPlane limitation \n		-	see manual for new settings option

STE-C 55X Autopilot/data/modules/Custom Module/preprocessor.lua
--[[
STE-C 55X Autopilot Plugin, a plugin to bring the "S-TEC FIFTY FIVE X" into your XPlane Flight Sim.
Copyright (C) 2022 Luka Emanuel Nöckel

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.

You can contact me via E-MAIL: quackderbruchpilot05@gmail.com, via GitHub (QuackDerBruchpilot) or the X-Plane.org forum (@flightsimmer1710)

	ANY USE OF THIS CODE CONSTITUTES ACCEPTANCE OF THE TERMS OF THE COPYRIGHT NOTICE.
	The current software uses the SASL library. Redistribution in source and binary forms, with 
	or without modification, are not permitted.
	Copyright (c) 2020, 1-sim.aero
	All rights reserved.
	SASL copyright holders are not responsible in any event for malfunction, unavailability, 
	incompatibility or other problems that might occur while using the end product which uses 
	SASL in conjunction with any other software. SASL copyright holders are not liable for any 
	properties for any end product which uses SASL.
]]--

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------- AP mode variables -------------------------------------------------------------------------------------------
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
sasl.logInfo("Preprocessor active, reloading...")

sasl.logInfo("STE-C 55X Autopilot Plugin - Copyright (C) 2022 Luka Emanuel Nöckel")
sasl.logInfo("This program comes with ABSOLUTELY NO WARRANTY. This is free software, and you are welcome to redistribute it under certain conditions; for details open the License file in the plugin's directory.")

apHDG = false
apNAV = false
apAPR = false
apREV = false
apALT = false
apVS = false

anGS = false
anUP = false
anDN = false
anGPSS = false

vsANN = false
altANN = false
rollMode = false
pitchMode = false
VSrate = 0
s = false
i = false
k = false
t = false
x = false
y = false
z = false

done = true
NavCount = 0
GPSS_override = false

an_FAIL = false																					-- WORK IN PROGRESS
an_CWS = false																					-- WORK IN PROGRESS
an_RDY = false																					-- WORK IN PROGRESS
																					


STE-C 55X Autopilot/data/modules/main.lua
--[[
STE-C 55X Autopilot Plugin, a plugin to bring the "S-TEC FIFTY FIVE X" into your XPlane Flight Sim.
Copyright (C) 2022 Luka Emanuel Nöckel

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

    	This program is distributed in the hope that it will be useful,
    	but WITHOUT ANY WARRANTY; without even the implied warranty of
		MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    	GNU General Public License for more details.

    	You should have received a copy of the GNU General Public License
    	along with this program.  If not, see <https://www.gnu.org/licenses/>.

You can contact me via E-MAIL: quackderbruchpilot05@gmail.com, via GitHub (QuackDerBruchpilot) or the X-Plane.org forum (@flightsimmer1710)

	ANY USE OF THIS CODE CONSTITUTES ACCEPTANCE OF THE TERMS OF THE COPYRIGHT NOTICE.
	The current software uses the SASL library. Redistribution in source and binary forms, with 
	or without modification, are not permitted.

	Copyright (c) 2020, 1-sim.aero
	All rights reserved.

	SASL copyright holders are not responsible in any event for malfunction, unavailability, 
	incompatibility or other problems that might occur while using the end product which uses 
	SASL in conjunction with any other software. SASL copyright holders are not liable for any 
	properties for any end product which uses SASL.
]]--

sasl.options.setAircraftPanelRendering(false) 
sasl.options.set3DRendering(false)
sasl.options.setRenderingMode2D(SASL_RENDER_2D_DEFAULT)
sasl.options.setInteractivity(true)


components = {
	autopilot {},
	settings {},
}

--------------------------------------------------------------------------------------------------- apWindow

apWindow = contextWindow {				-- create new window

	name = "STE-C 55X Autopilot";				-- title of window
	position = {0, 0, 783, 202};				-- spawning position and size of window
	visible = false;							-- visible on load
	gravity = {0, 1, 0, 1};						-- no changes when XPWindow changes
	noResize = true;							-- no Resizing allowed
	noDecore = true;							-- no Decore
	
	components = {								
		autopilot {								-- content of window 
			position = {0, 0, 783, 202};
		};
	};
}

function show_hide_ap()					-- handle to set/unset window visible whenever we want
	apWindow:setIsVisible(not apWindow:isVisible())
end

--------------------------------------------------------------------------------------------------- settings window

settings_Window = contextWindow {		-- create new window

	name = "STE-C 55X Autopilot - Settings";	-- title of window
	position = {100, 150, 400, 300};			-- spawning position and size of window
	visible = false;							-- visible on load
	gravity = {0, 1, 0, 1};						-- no changes when XPWindow changes
	noResize = true;							-- no resizing allowed
	noDecore = false;							-- window has decore
	
	components = {
		settings {								-- content of window
			position = {0, 0, 400, 300};
		};
	};
}


function show_hide_set()				-- handle to set/unset window visible whenever we want
	settings_Window:setIsVisible(not settings_Window:isVisible())
end

--------------------------------------------------------------------------------------------------------------------------------- configuring menu

menu_master	= sasl.appendMenuItem (				--
	PLUGINS_MENU_ID, "STE-C 55X Autopilot"		-- add new Item to plugin menu list
)												--

menu_main	= sasl.createMenu (					--
	"", PLUGINS_MENU_ID, menu_master			-- add menu to our entry in plugin menu list
)												--

menu_ap	= sasl.appendMenuItem (					--
	menu_main, "show/hide", show_hide_ap		-- add option to our plugin menu Item (show/hide AP device)
)												--

menu_settings = sasl.appendMenuItem (			--
	menu_main, "settings", show_hide_set		-- add option to our plugin menu Item (show settings)
)												--

--------------------------------------------------------------------------------------------------------------------------------- include settings from config

addSearchPath(moduleDirectory.."/Custom Module/")				--
include("config")												--
																-- load config file and apply them
if AP_wnd_on_startup then										--
	apWindow:setIsVisible(true)									--
end																--

if first_launch then											--
	settings_Window:setIsVisible(true)							-- open settings Window on first launch
end																--